components/script/dom/htmlinputelement.rs

```diff
@@ -1135,14 +1263,46 @@ impl HTMLInputElement {
         let shadow_tree = self.shadow_tree.borrow();
         Ref::filter_map(shadow_tree, |shadow_tree| {
             let shadow_tree = shadow_tree.as_ref()?;
-            let ShadowTree::Color(color_tree) = shadow_tree;
-            Some(color_tree)
+            match shadow_tree {
+                ShadowTree::Color(color_tree) => Some(color_tree),
+                _ => None,
+            }
         })
         .ok()
         .expect("UA shadow tree was not created")
     }
 
     fn update_shadow_tree_if_needed(&self, can_gc: CanGc) {
+        if self.input_type() == InputType::Text {
+            let text_shadow_tree = self.text_shadow_tree(can_gc);
+            let value = self.Value();
+
+            let placeholder_text = match value.is_empty() {
+                true => self.placeholder.to_owned().take(),
+                false => DOMString::new(),
+            };
+
+            // The addition of zero-width space here forces the text input to have an inline formatting
+            // context that might otherwise be trimmed if there's no text. This is important to ensure
+            // that the input element is at least as tall as the line gap of the caret:
+            // <https://drafts.csswg.org/css-ui/#element-with-default-preferred-size>.
+            //
+            // This is also used to ensure that the caret will still be rendered when the input is empty.
+            // TODO: Is there a less hacky way to do this?
+            let value_text = match value.is_empty() {
+                false => value,
+                true => "\u{200B}".into(),
```

Comment:
A linebreak (`\n`) is enough.

Reply:
I found that line break `\n` wouldn't trigger caret to render right now.

Reply:
So in Servo `<div><br></div>` has zero height? That seems like a bug.